### DISCUSSION TITLE
ArrayPool<T>/MemoryPool<T> Best Practice

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/runtime#48257] [General]
When should they be used? Are there any guidelines?

And the docs said:

> Using the ArrayPool<T> class to rent and return buffers (using the Rent and Return methods) can improve performance in situations where arrays are created and destroyed frequently, resulting in significant memory pressure on the garbage collector.

It's not very clear, how frequent is frequent?

Is it premature optimization to use them instead of temporary arrays which created in the all method bodies? 
Even refactor all short life object like the follow, is this bad?
```cs
class ShortLife : IDisposable {
    private byte[] _buffer;
    
    public ShortLife(...) {
        _buffer = ArrayPool<byte>.Shared.Rent(xxx);
    }
    public void Dispose() {
        ArrayPool<byte>.Shared.Return(_buffer);
    }
}

// then use
using var shortLife = new ShortLife(...);
```

**Comment by davidfowl** [comment]
The shared array pool is optimized for synchronous rent and return. I think that's where the short lived advice comes from.

**Comment by Clockwork-Muse** [comment]
The other thing that should be mentioned here is that if your array is also small (in bytes used - size of type * number of elements), in addition to being short-lived, it may behoove you to look at using `stackalloc`ed arrays.

**Reply by GF-Huang** [reply]
In the async method body, `stackalloc` seems cumbersome to use and requires a static local function.

**Reply by Clockwork-Muse** [reply]
And that's just another part of the determination.  Although note that if you're in an `async` method, they may not "live" long enough - since in many cases you use the array as a buffer to an async call, you'd run into problems at that point.

**Comment by akeit0** [comment]
I think we need guidelines.

I was optimizing Linq's TakeLast, but I have a different opinion from the reviewer regarding ArrayPool and yield. I would like official guidelines (at least for implementations within the BCL).


Me:
https://github.com/dotnet/runtime/pull/118490#issuecomment-3165782684
> The reason I return the array just before yielding is that not all yielding are done synchronously, and it is possible to hold an array rent from ArrayPool. In this case, it is customary not to use rent arrays.

Reviewer:
https://github.com/dotnet/runtime/pull/118490#discussion_r2390094571
> Pooled buffers can certainly cross yield boundaries, it's just that you can no longer use stack allocated buffers.

**Comment by ichensky** [comment]
The use of ArrayPool<T>.Rent and Return can be used, for example, within a loop where an array of varying size needs to be created multiple times. The ArrayPool attempts to re-use memory allocated for prior arrays.

```csharp
for (int i =0; ... ){
    var buffer = ArrayPool<int>.Shared.Rent(N);
    ...
    ArrayPool.Return(buffer, clearArray: true);
}
```
However, in many cases, when N is constant (or when the exact value doesn't strictly matter), such code can be optimized by moving the buffer to an outer loop or function and just re-using it.

```csharp
int[] buffer = new int[N];
for (int i =0; ... ){
    ...
}
```

**Reply by Clockwork-Muse** [reply]
Generally, you'd still rent/return even in the outer context.
Additionally, you're (almost) always supposed to use `try { ... } finally { ... }` to ensure that arrays are returned in the event of exceptions.

**Reply by gfoidl** [reply]
> Additionally, you're (almost) always supposed to use `try { ... } finally { ... }` to ensure that arrays are returned in the event of exceptions.

No.
That can yield in use-after-retrun bug which are hard to diagnose. Now there's even an [official best practice](https://learn.microsoft.com/en-us/dotnet/standard/unsafe-code/best-practices#20-arraypooltshared-and-similar-pooling-apis) against using try-finally.

In the event of exceptions just let the GC collect the (rented) array. The array pool can handle such cases w/o problems.

**Reply by Clockwork-Muse** [reply]
.... and here I was approaching it from the perspective of it being the only return in the method.

**Reply by martin-steinhoefer** [reply]
Hi,

@gfoidl can you please bring an example of such a case where using a `try { ... } finally { ... }` leads to a use-after-return bug?

**Reply by bartonjs** [reply]
```C#
private async Task<Something> DoStuffAsync(CancellationToken cancellationToken) {
    byte[] pooled = ArrayPool<byte>.Shared.Rent(4096);

    try {
        _ = DoStuffAsyncCore(pooled);
        await SomeOtherSignal(cancellationToken);
        return ExtractAnswer(pooled);
    } catch {
        ArrayPool<byte>.Shared.Return(pooled);
    }
}
```

This sort of extreme hopefully doesn't directly happen, but if SomeOtherSignal gets interrupted by the cancellation token, nothing says DoStuffAsyncCore has finished writing to the array.

A realistic version of the above could happen in I/O, if a native async operation is cancelled while a write is in progress.  The cancelling thread may be informed the cancellation happened (really "was recorded"), and can then move into the rest of the finally block, but the multi-threaded operation in progress is already beyond checking for the token.

Too convoluted?  OK, well, what about something like

```C#
private async Task<int> Worker(byte[] buffer, CancellationToken cancellationToken) {
  int written = await GoFetchALargeThingAsync(buffer, cancellationToken);

  // Cancellation happens here.  The whole of, or at least part of, the reverse is done after cancellation.
  // The thread that did the cancelling has moved on, maybe rented the array back.

  buffer.AsSpan(0, written).Reverse();
  return written;
}
```

**Reply by EgorBo** [reply]
I think developers are just scared that not retuning an array to the pool will lead to potential memory leaks which is generally not true, so they wrap with try-finally to be leaks-proof.

**Reply by martin-steinhoefer** [reply]
Thank you for your response. 

The example violates ownership constraints (pooled escapes). That would be the case even **without** the `try { ... } finally { ... }` construct. That construct is unable to help in such usage scenarios. 

I should have asked for a example, where placing the `Return` into a `finally` block actually causes a use-after-free problem - I couldn't come up with one.

The guidance mentioned is somewhat misleading to me: 
It tells the reader that it is bad practice to use `try { ... } finally { ... }` constructs because it does not save from the escape problem. Although that is true - it completely misses the point:
In case of `ArrayPool<T>` the only resource managed is memory - which ich actually managed by GC anyway. So missing to `Return` a rented array is no problem. So the pattern is **not required** but using it should **not** be dis-couraged. In fact, the guidance additionally refers to `ObjectPool<T>` as a similar case - where following it actually **can** create problems - if `T` holds resources other than memory (f.i. `SqlConnection`). Following the guidance in this case will lead to problems.
The guidance should have been like:
DO keep matching....
DON'T use `try/finally` pattern because it complicates source and adds no value. Arrays obtained by `Rent` are perfectly cleaned up by GC.
DO be aware, that this does apply **only** to arrays obtained by `ArrayPool<T>.Shared.Rent` (and does **not** apply to `ObjectPool<T>`)

**Reply by EgorBo** [reply]
@martin-steinhoefer 

> DON'T use try/finally pattern because it complicates source and adds no value. 

The example above perfectly shows why try-finally may cause use-after-free where just lack of try-finally doesn't have an issue, it's definitely not just "adds no value.". It's indeed a general ownership/lifetime issue that cannot be solved today in modern C#, so at very least we don't recommend certain patterns

**Reply by martin-steinhoefer** [reply]
It's the very same as with ownership of `IDisposable` instances. If the instance escapes - it **MUST NOT** be released (returned rsp. disposed). 

Usage of 'Return' (or 'Dispose') is contra-indicated if the instance escapes. That is, what the guideline should state - **clearly**. 
And usage of the pattern in a valid scenario adds little value - as abandoned buffers are perfectly fine GC-ed. That is, what the guideline should state.
And the guideline should **not** refer to `ObjectPool<T>` as completely different thoughts apply.

**Reply by EgorBo** [reply]
> it MUST NOT be released (returned rsp. disposed).

That is not a correct guideline I think, some types might be perfectly protected from use-after-free (by checking isDisposed internally) and can lead to a bad behavior if not disposed at all.

> And the guideline should not refer to ObjectPool<T> as completely different thoughts apply.

No, same rules apply to ObjectPool:
```cs
var pool = ObjectPool.Create<StringBuilder>();
var obj = pool.Get();
pool.Return(obj);
var obj2 = pool.Get();

obj.Append("hello"); // use-after-free
Console.WriteLine(obj2.ToString()); // prints "hello"
```

**Reply by martin-steinhoefer** [reply]
The fact, that both `ArrayPool<T>` and `ObjectPool<T>` are prone to use-after-free bugs is undisputed.

In your `ObjectPool<T>` example you missed the fact, that `T` itself can be a resource (`IDisposable`) and missing to dispose it can lead to resource leak. Contrary to `ArrayPool<T>` which gets you merely memory (the array contains nothing - you put **your** objects into it) on `ObjectPool<T>` you get **an instance** - which you get ownership of - so you are obliged to release it. You can take care of your responsibility by either invoking `Return` or `Dispose` - but you are obliged to do either (if the instance does not escape - otherwise the caller becomes the owner and responsible for invoking `Dispose`). You are freed from this **only** if `T` itself does not need releasing (contrary to `ArrayPool<T>`). If you did take care of your responsibility, accessing the instance afterwards may result in a use-after-free bug.

Why is the the rule, that if the instance escapes then it **must not** be released, a bad guidance? I did not say, that those must not be released (disposed) at all. 
Ownership transfers to the caller / receiver which then becomes responsible for releasing the object - in case of `ArrayPool<T>` that is easy - just let go of the reference. In case of `IDisposable` you become responsible for invoking `Dispose`. However, releasing the resource  early if the instance escapes is likely an error in both cases. Consider the usual `IDisposable` implementation - when diposed most if the members throw `ObjectDisposedException` - so releasing (disposing) an escaped instance will likely lead to throwing (luckily we usually get an exception here - contrary to the hard to detect `ArrayPool<T>` case). 
Just because you can imagine types, that are perfectly protected against use-after-free - most of them aren't. And a guideline should target the common case - while not hindering the special one. 

Perhaps the guideline should be more detailed and explain the background (maybe using examples).

**Reply by martin-steinhoefer** [reply]
The `try/finally` pattern is all about preventing memory/resource leaks in case of exceptions. But it does not come for free. using the pattern has it's drawbacks. One of them is it increases the likelihood of use-after-free bugs. Another is, that it prevents inlining.
The pattern is **not** about preventing or introducing use-after-free bugs. Applying it does not automatically introduce use-after-free bugs. But as it increases complexity of program flow, the possibility of any bug is increased. However it can be implemented correctly - without causing such conditions.

Given the actual implementation of `ArrayPool<T>`  (and perhaps `MemoryPool<T>`), that failing to return a rented instance does **not** introduce a memory/resource leak, makes the pattern **not applicable** as there is no leak to prevent from. So the guideline should state:

- DON'T use the `try/finally` pattern to ensure, instances rented from `ArrayPool<T>` (`MemoryPool<T>`) are returned to the pool. The pattern is not applicable, because failing to return the instance does not cause a memory/resource leak. Using the pattern needlessly increases the complexity of program flow and the likelihood of errors such as use-after-free conditions.

Concerning `ObjectPool<T>` the situation is somewhat different. As it returns **instances** of `T` (as opposed to 'containers' of `T`), there is no general knowledge, whether failing to return an instance would cause a memory/resource leak. One could have pooled `StringBuilder` where it poses no problem, but one also may have pooled `IDisposable` where it certainly **is** a problem. The guideline should state:

- DO use the `try/finally` pattern to ensure, instances rented from `ObjectPool<T>` are returned to the pool - unless you are certain, that failing to return it, poses no risk of memory/resource leak. When using the pattern, be aware that its increased complexity makes it more likely to have other errors (like use-after-free conditions).

It's 'DO' and not 'CONSIDER' (and certainly **not** 'AVOID' or 'DON'T' ) - because the guideline should provide safe solutions (for the less experienced programmer) in the general case. Exceptions are clearly marked - for the case the programmer can induce additional knowledge. But that's 'opt-out'. Additionally, the guideline notes, that using the pattern has it's own drawbacks - to make the programmer aware of common pitfalls.

Contrary to failing to return the instance being no problem - returning it more than once actually **is** introducing a use-after-free bug. 
So the guideline should state:

- DON'T return an instance rented from either `ArrayPool<T>`, `MemoryPool<T>` or `ObjectPool<T>` more than once to the pool. This also applies to different pool instances. Doing so results in hard to detect use-after-free bugs.

**Reply by martin-steinhoefer** [reply]
> ```cs
> private async Task<Something> DoStuffAsync(CancellationToken cancellationToken) {
>     byte[] pooled = ArrayPool<byte>.Shared.Rent(4096);
> 
>     try {
>         _ = DoStuffAsyncCore(pooled);
>         await SomeOtherSignal(cancellationToken);
>         return ExtractAnswer(pooled);
>     } catch {
>         ArrayPool<byte>.Shared.Return(pooled);
>     }
> }
> ```
> 
@bartonjs 
Hi, 
I think the example you provided has a serious problem. You call `DoStuffAsyncCore` and `ExtractAnswer`- which seems to be third party code or code that invokes something asynchron. Anyway, you provided no insights about the code. So the programmer seems **not** able to infer, what's going on in those methods. In the general case, if one is unable to tell what's happening in the code, one should assume it being malicious - that is 
- it stores the instance elsewhere to refer to it later
- it shares the instance to other threads
- it invokes code that runs delayed or otherwise asynchronous
- it invokes `Return` itself on any matching pool
Any of those will cause the invocation of `Return` to cause use-after-free conditions. As you can't know what gets called, you can't know whether you safely can invoke `Return`. If you can't know that, then you simply could omit the invocation - but that would violate the first guideline ('DO keep matched calls to `Rent` and `Return`...').

If you can't fulfill even the first rule, then perhaps usage of `ArrayPool<T>` ist not indicated.

BTW: If `DoStuffAsyncCore` invokes `Return` with the provided instance on **any** instance of `ArrayPool<byte>`, then even the invocation of `ExtractAnswer` would have use-after-free conditions. If one can't exclude such things, usage of `ArrayPool` seems not appropriate.
As a rule of thumb - only if one can reason about the **whole** call graph and can exclude 'escape conditions' like above - only then usage of `ArrayPool<T>` is appropriate. And the more code gets called in that section, the more likely it gets that the instance escapes.

I think, the guidelines should clearly state that too.

PS: your example doesn't even have the `try/finally` pattern in it.

**Reply by bartonjs** [reply]
Any async method, almost by definition, stores the parameter to refer to it later.  In the example you quoted, I'm showing a try/finally (which apparently I miswrote as try/catch) that would have a problem.  DoStuffAsyncCore isn't subscribed on the same cancellation token, perhaps because it's doing native interop, like

```C#
private unsafe void DoStuffAsyncCore(byte[] buffer) {
  Interop.SomeLib.DoStuff(_context, buffer, buffer.Length, out int bytesWritten);
  Notify(bytesWritten);
}
```

In the normal flow, the task from DoStuffAsync doesn't return the rented buffer until the notify happens, which unblocks the task in ExtractAnswer.  But, if cancellation is invoked, ExtractAnswer will abort with an OperationCancelledException, then DoStuffAsync will return the buffer to the pool, and now some other caller can get that array back from the next call to Rent.  Right after that rent, the native operation finishes doing whatever made it take so long, and it now does a UAF write into the buffer.

So, this operation clearly has a potential UAF.  "OK, so don't use try/finally here".  Sure.  Now change it so DoStuffAsync takes a buffer, but handles it in the exact same way.  The caller of DoStuffAsync has no way of ever knowing that DoStuffAsync has a potential write-after-cancel, so it uses a try/finally... and now the UAF write is still present.  There's nowhere in docs that we do or don't guarantee safety with regards to pooled arrays as input and exception flows for asynchronous operations.  We _try_ to be tidy about it, but don't guarantee it.  Anything with a virtual (e.g. Stream) is unprovable, since it could be an unknown Stream.

Technically speaking, the non-throwing form isn't "safe", either, but it's pretty "expected" that an async operation will not hold the reference beyond successful completion.  "Unsuccessful completion" is harder to reason about, particularly if Task shards out into parallel tasks.  And that's why our guidance is to not use try/finally with rent/return in async.  You'll see try/finally rent/return done in a few places, but they are probably always locally-synchronous code, and not calling into anything making use of `virtual` (e.g. it gets used in some crypto code wrapping a P/Invoke, because we know that once any exception (or return) happens the native code is no longer going to be writing to (or reading from) that segment of memory)

**Reply by martin-steinhoefer** [reply]
I do understand your arguments like that:

Given you have async code that:
- uses a pooled array 
- runs synchronously in the regular case
- returns early in case of exception/cancellation

Then application of the `try/finally` pattern to return the rented array introduces a UAF condition (in the exceptional case) - while not applying the pattern wouldn't have the UAF condition.

That seems to be one of the not-so-obvious pitfalls of using the pattern. Actually, that makes the pattern not applicable to such code in general. The correct solution would be returning the rented resource in the correct continuation - that is surely not a `finally` block around the code in question.

That makes it harder to reason about the `ObjectPool<T>` scenario. Because my argument about `T` itself holding resources still holds. Perhaps the guideline should state that the pattern might not be applicable to such code. Instead a more complex solution must be taken.

My argument, that the pattern is not applicable in the `ArrayPool<T>` scenario still holds.

**Comment by martin-steinhoefer** [comment]
> When should they be used? Are there any guidelines?
> 
> And the docs said:
> 
> > Using the ArrayPool class to rent and return buffers (using the Rent and Return methods) can improve performance in situations where arrays are created and destroyed frequently, resulting in significant memory pressure on the garbage collector.
> 
> It's not very clear, how frequent is frequent?
> 
> Is it premature optimization to use them instead of temporary arrays which created in the all method bodies? Even refactor all short life object like the follow, is this bad?
> 
> ```cs
> class ShortLife : IDisposable {
>     private byte[] _buffer;
>     
>     public ShortLife(...) {
>         _buffer = ArrayPool<byte>.Shared.Rent(xxx);
>     }
>     public void Dispose() {
>         ArrayPool<byte>.Shared.Return(_buffer);
>     }
> }
> 
> // then use
> using var shortLife = new ShortLife(...);
> ```

That is actually trading creation of an `IDisposable` for creation of an array instance. Only in case the array to be rented is large enough that will pay off. Usage of pooled arrays seems not to benefit from wrapping them up in `IDisposable` (pooled objects may be a different story):  - You want the array - the `IDisposable` is **not** an array (and making it mimicking one is cumbersome).
